Fixes #39

mondracek opened the pull request:
Fixes #39 

This should address issue #39 (automatic loading of FHI-AIMS geometry.in and VAS POSCAR files). Now, you can do e.g.
`ppafm-generate-ljff -i geometry.in`
`ppafm-generate-ljff -i poscar`
or
`ppafm-generate-ljff -i vasp_file_with_a_funny_name --format poscar`
and the `ppafm` code recognizes the input geometry format. Valid options for the `format` argument are `xyz`, `in`, `poscar`, `xsf`, and `cube`.

We should probably add some automatic tests for this functionality but I have not come up with anything yet. Obviously, the natural way is to create some example files in the `geometry.in`, `atoms.xyz` and `POSCAR` formats and see if they load correctly. But to include this into automatic testing, we would essentially need to create new example folders and distribute them with the code, right? Do we want to do it just to have some tests available?

NikoOinonen commented on ppafm/io.py:
```diff
@@ -374,15 +389,19 @@ def loadGeometry(fname=None,params=None):
+        atoms, nDim, lvec = loadGeometryIN(fname)
+    #TODO: introduce a function which reads the geometry from the .npy file
     else:
         sys.exit("ERROR!!! Unknown format of geometry system. Supported formats are: .xyz, .cube, .xsf \n\n")
```
The message is wrong (we support more formats), and it would be better if this was a `raise ValueError()` instead of `sys.exit()`, so that if you use this function in some place that can catch the error, it won't just suddenly exit the whole program.

NikoOinonen commented on ppafm/cli/generateDFTD3.py:
```diff
@@ -22,6 +22,7 @@ def main():
 
     parser.add_argument('-i', '--input',       action='store',
         help='Input file. Mandatory. Supported formats are: .xyz, .cube, .xsf.')
+    parser.add_argument("--format", action="store", type="string", help="Format of the input geometry file (overrides format concluded from the file name extension)", default=None)
     parser.add_argument('-f', '--data_format', action='store', default='xsf',
```
I think it would better to reconsider the names of the arguments. The `format` and `data_format` may be confusing for a new user. Maybe `input_format` and `output_format` instead.

But I think the CLI arguments need an overhaul anyway, so this could be done separately. Also related to #148.

NikoOinonen replied:
I'm working on overhauling the CLI arguments as a part of resolving #148, so let's not touch this here.

NikoOinonen reviewed (changes requested):
A couple of things to fix, otherwise looks good.

NikoOinonen commented:
> But to include this into automatic testing, we would essentially need to create new example folders and distribute them with the code, right? Do we want to do it just to have some tests available?

We have a dedicated folder called `tests` in the root of the repo. This folder is not distributed with the rest of the code with pip, so we are free to add anything we want there and it won't interfere with the users. I don't think it would be an issue to add a couple of example input files there (maybe in a subfolder), since these file formats are very small in size.

mondracek commented:
The typo and the error messages pointed out by @NikoOinonen were addressed in the last commit(s). Niko, I will now wait till you resolve #148, so that we can decide what the name of the argument for input format specification shall actually be ("input_format" perhaps?).

mondracek commented:
The name of the argument that specifies the input data format is now (after merging work done in PR #185)
`--input_format`, short version `-F`.
E.g.
```ppafm-generate-ljff -i geometry.out -F in```
or
```ppafm-generate-elff -i LOCPOT.xsf~ --input_format xsf```
etc.
The argument is not case sensitive (`XSF` and `xsf` would be the same).

While integrating this into `generateElFF`, I stumbled over how the format of some other input files, those passed through the `--tip_dens`, `--KPFM_tip`, and `--KPFM_sample` arguments, should be treated. The latter two files may be needed for KPFM, the first would be a tip charge density  from _first principles_, not necessarily for KPFM only. As for now, the format of these files is determined based on the file name extension, which I guess should stay that way, no need to add an analogy of the `--input_format` for each of them. I have just added the possibility of `tip_dens` to be in the `CUBE` format and, furthermore, I have added checks that all these files are either `*.cube` or `*.xsf` and that the `KPFM_tip` and , `KPFM_sample` are both in the same format as the main input file (specified by `-i`) , otherwise an error is issued. But as the KPFM is not working at the moment, these additions remain untested.

mondracek commented:
**Trouble with `flake8` when committing**
I have made the last commit with `git commit -n`, that is, with the automatic tests turned off. Unless I do this, I get *lots* of complaints from `flake8` about various formatting issues and the test fails. I actually find most of the formatting suggestions by `flake8` reasonable, but
- there are really many of them and
- most of them are not even about the code I have written myself as a part of this specific branch/PR,

so I do not think this here is the right place to implement them. I can start a new issue and split off a dedicated branch, if needed. Anyway, I wonder, am I the only one who hit into this issue? Is everybody making *commits* with the `-no-verify` option?

NikoOinonen commented:
>  Is everybody making commits with the -no-verify option?

Probably yes, because there are _a lot of_ problems in the current code that flake8 does not approve of. Actually most of these are code style things, but there is no consensus on whether we want to enforce a style or not. For now you just have to use the `--no-verify` option. (Actually, we probably should disable the flake8 test until there is some agreement on it...)

NikoOinonen commented:
@mondracek Is this ready for review? If it is, you can re-request the review from the right side using the circling arrows button next to my user name.

NikoOinonen commented on ppafm/cli/generateElFF.py:
```diff
@@ -77,14 +88,18 @@ def main():
+        print(">>> Loading tip density from ",args.tip_dens,"...")
+        if(args.tip_dens.lowercase().endswith("xsf") ):
+            rho_tip, lvec_tip, nDim_tip, head_tip = io.loadXSF( args.tip_dens )
+        elif(args.tip_dens.lowercase().endswith("cube") ):
+            rho_tip, lvec_tip, nDim_tip, head_tip = io.loadCUBE( args.tip_dens )
```
Actually, this doesn't work, because the `loadCUBE` function automatically multiplies the array by the conversion factor between Hartree and eV, but here we have an electron density instead of a Hartree potential. I would remove this for now. If we want to support .cube also for electron densities, we need to fix the problem with the factor, which is probably another issue.

NikoOinonen reviewed (approved):
Ok, looks good now.
